### DISCUSSION TITLE
Invocable Actions REST API endpoint support

### DISCUSSION THREAD

**Discussion by stomita** [jsforce/jsforce#1440] [Ideas]
From Winter'15
https://developer.salesforce.com/docs/atlas.en-us.192.0.api_rest.meta/api_rest/resources_actions_invocable.htm

**Comment by stomita** [comment]
Changed milestone because of no such good usage senario until 33.0 API version, in which custom action will be available.

**Comment by DougMidgley** [comment]
Any Updates on this?

**Comment by xialisun** [comment]
we are also using this on a node.js service to call into a InvocableMethod as a rest web service. Any update on when this feature will be arriving?

**Comment by jesusbarcenilla** [comment]
Until there is an official solution here is an example to create a custom notification but keep using jsforce connection to handle most of the tedious work of calling the API:

conn.requestPost('/actions/standard/customNotificationAction', { "inputs" :[{
  title: "JSForce Custom Notification",
  body: "The notifications are coming from JSForce!",
  customNotifTypeId: '0ML1F0000004IOYWA2',
  targetId: '0061F00000DXxX2QAL',
  recipientIds: ['0051F00000nWKatQAG']
}]}, undefined, console.log)

I think It easy to adapt it to any other use of the actions API, event extending jsforce prototype.

**Comment by nmandyam** [comment]
@jesusbarcenilla Any idea if this continues to work now? Because the request gives me a 404 and a message: "Invalid Action Type: customNotificationAction".
@stomita Is there another way to send notifications now?

**Comment by jesusbarcenilla** [comment]
@nmandyam That code was part of a spike and I think we ended using triggers but It shouldn't be any reason for that code to stop working. 

I'd check:

- if the custom notification type was created and the id match the customNotifTypeId
- if the user has the send notification permission 
- if the final url the browers use is the right one
You can find a bit more info (not a lot) here https://developer.salesforce.com/docs/atlas.en-us.api_action.meta/api_action/actions_obj_custom_notification.htm

**Comment by nmandyam** [comment]
Thanks. Unfortunately, I could not figure out what was wrong - I did check the items you listed before I posted. OTOH, I did get this working, using jsForce's Anonymous Apex capability:
`var apexBody = "String targetId = '<rec target ID>'; Set<string> rIds= new Set<string>(); rIds.add('<recipient-user-id'); Messaging.CustomNotification notification = new Messaging.CustomNotification(); notification.setTitle('Apex Custom Notification'); notification.setBody('The notifications are coming from INSIDE the Apex!'); notification.setNotificationTypeId('<custom-notification-type-id>'); notification.setTargetId(targetId); try { notification.send(rIds); } catch (Exception e) { System.debug('Problem sending notification: ' + e.getMessage()); }";`
` `
`conn.tooling.executeAnonymous(apexBody, function(err, res) {`
`  if (err) { return console.error(err); }`
`});`

BTW, I got the custom-notification-type-id using this SOQL: `SELECT Id, DeveloperName FROM CustomNotificationType WHERE DeveloperName='<notification-type-api-name>'`

Thanks again for setting me on the right path!

**Comment by cydharttha** [comment]
I was able to get the above detailed `requestPost` method to work by setting the API version for the jsforce connection:

```
const sfConnection = new jsforce.Connection({ version: '56.0' })
```

Without this, the connection was being created with an API version of 42.0, which is lower than v46.0 where `customNotificationAction` became available.

I used tooling to find the `CustomNotificationType` Id by DeveloperName:

```
const notificationType = await sfConnection.tooling
  .sobject('CustomNotificationType')
  .find({ DeveloperName: 'DocExchange' })
  .execute()
```

Thanks!!
